File: 39.组合总和/combinationSum.go
package problem0039

import "sort"

func combinationSum(candidates []int, target int) [][]int {
	var (
		result    [][]int
		backTrack func(path []int, target int, index int) (shouldGoOn bool)
	)
	backTrack = func(path []int, target int, index int) (shouldGoOn bool) {
		if target == 0 {
			temp := make([]int, len(path))
			copy(temp, path)
			result = append(result, temp)
			return true
		}
		if target < 0 {
			return false
		}
		for i := index; i < len(candidates); i++ {
			path = append(path, candidates[i])
			if !backTrack(path, target-candidates[i], i) {
				break
			}
			path = path[:len(path)-1]
		}
		return true
	}
	sort.Ints(candidates)
	backTrack([]int{}, target, 0)
	return result
}

File: 39.组合总和/combinationSum.js
function combinationSum(candidates, target) {
	const result = []
	const N = candidates.length

	const backTrack = (combine, targetSum, index) => {
		if (index === N) {
			return
		}
		if (targetSum === 0) {
			result.push(combine)
			return
		}
		// 不选择当前数
		backTrack(combine, targetSum, index + 1)
		// 选择当前数
		if (targetSum - candidates[index] >= 0) {
			combine.push(candidates[index])
			backTrack([...combine], targetSum - candidates[index], index)
			combine.pop()
		}
	}

	backTrack([], target, 0)

	return result
}